course6/toyapp-adam/math/vector.h

```diff
@@ -1,113 +1,172 @@
+		return true;
+	}
+
+	friend bool operator!=(const Vector& lhs, const Vector& rhs)
+	{
+		return !operator==(lhs, rhs);
+	}
+
+	Vector& operator=(const Vector& other)
+	{
+		for (unsigned i = 0; i < Dimensions; ++i)
+			mComponents[i] = other[i];
 		return *this;
 	}
 
-	Vector2D& operator-=(const Vector2D& rhs)
+	Vector& operator+=(const Vector& other)
 	{
-		x -= rhs.x;
-		y -= rhs.y;
+		for (unsigned i = 0; i < Dimensions; ++i)
+			mComponents[i] += other[i];
 		return *this;
 	}
 
-	Vector2D& operator*=(float scalar)
+	Vector& operator*=(FloatType scalar)
 	{
-		x *= scalar;
-		y *= scalar;
+		for (unsigned i = 0; i < Dimensions; ++i)
+			mComponents[i] *= scalar;
 		return *this;
 	}
 
-};
+	Vector& operator-=(const Vector& other)
+	{
+		return operator+=(-other);
+	}
 
-struct Vector3D
-{
-	float x = 0.0f;
-	float y = 0.0f;
-	float z = 0.0f;
+	Type& operator[](unsigned index)
+	{
+		assert(index < Dimensions);
+		return mComponents[index];
+	}
 
-	Vector3D operator-() const { return {-x, -y, -z}; }
-	Vector3D operator+(const Vector3D& rhs) const { return {x + rhs.x, y + rhs.y, z + rhs.z}; }
-	Vector3D operator-(const Vector3D& rhs) const { return {x - rhs.x, y - rhs.y, z - rhs.z}; }
-	Vector3D operator*(float scalar) const { return {x * scalar, y * scalar, z * scalar}; }
+	Type operator[](unsigned index) const
+	{
+		assert(index < Dimensions);
+		return mComponents[index];
+	}
 
-	Vector3D& operator+=(const Vector3D& rhs)
+	Vector operator-() const
 	{
-		x += rhs.x;
-		y += rhs.y;
-		z += rhs.z;
-		return *this;
+		Vector result;
+		for (unsigned i = 0; i < Dimensions; ++i)
+			result[i] = -mComponents[i];
+		return result;
 	}
 
-	Vector3D& operator-=(const Vector3D& rhs)
+	Vector operator+(const Vector& rhs) const
 	{
-		x -= rhs.x;
-		y -= rhs.y;
-		z -= rhs.z;
-		return *this;
+		return Vector{*this} += rhs;
 	}
 
-	Vector3D& operator*=(float scalar)
+	Vector operator*(FloatType scalar) const
 	{
-		x *= scalar;
-		y *= scalar;
-		z *= scalar;
-		return *this;
+		return Vector{*this} *= scalar;
 	}
 
-	Vector3D Normalized() const
+	Vector operator-(const Vector& rhs) const
 	{
-		auto inverseLength = 1.0f / Length();
-		return {
-			x * inverseLength,
-			y * inverseLength,
-			z * inverseLength
-		};
-	};
+		return Vector{*this} -= rhs;
+	}
+
+	Vector Normalized() const
+	{
+		auto length = Length();
+		return !ApproximatelyEquals(length, 0.0)
+				? Vector{*this} *= 1.0 / length
+				: Vector{Type{1.0}, Type{0.0}};
+	}
+
+	FloatType Dot(const Vector& rhs) const
+	{
+		FloatType result{0.0};
+		for (unsigned i = 0; i < Dimensions; ++i)
+			result += mComponents[i] * rhs[i];
+		return result;
+	}
+
+	FloatType Length() const
+	{
+		return std::sqrt(Dot(*this));
+	}
 
-	Vector3D Cross(const Vector3D& rhs) const
+	friend std::ostream& operator<<(std::ostream& stream, const Vector& vector)
 	{
-		return {
-			y * rhs.z - z * rhs.y,
-			z * rhs.x - x * rhs.z,
-			x * rhs.y - y * rhs.x
-		};
+		stream << vector[0] << ',' << vector[1];
```

Comment:
Bug! won't work for vector with size != 2

Reply:
oops, oversight!